### PR TITLE
Updating regex tertiary methods

### PR DESCRIPTION
**Pull request by slnguyen**
This PR addresses #17226 and does the following:

- Replaces `string.search` and `bytes.search` (returns a regexMatch) with 'string.find' and `bytes.find`, which returns a 'byteIndex' that represents the offset where a regular expression and string/byte match
- Replaces 'string.matches' and `bytes.matches` (returns a regexMatch) with `string.startsWith` and `bytes.startsWith`, which returns true if a string/byte starts with a given regular expression
- Deprecates `string.matches` and `bytes.matches`
- Renames the `pattern` argument to `sep` in `string.split` and `bytes.split`
- Updates current tests and adds deprecation tests to reflect the above changes
- Adds documentation to reflect the above changes

### CODE REVIEW

File: modules/standard/Regex.chpl [hunk 1]
```diff
@@ -1153,28 +1153,52 @@ inline operator :(x: bytes, type t: regex(bytes)) throws {
   return compile(x);
 }
 
+/* Search the receiving string for a regular expression already compiled
+   by calling :proc:`regex.search`. Search for matches at any offset.
```
Comment (lydia-duncan) [inline modules/standard/Regex.chpl:1157]: I think this might go a little too into implementation details.  Maybe: `Search the receiving string for the result of a compiled regular expression.  Search for matches at any offset.` ?

It looks like you got this from other documentation in the file.  I think the same applies to those as well, but it's okay if you don't fix them in this PR.  Since you're adding these paragraphs, I do think it would be nice to avoid the implementation details, though.

Comment (lydia-duncan) [reply modules/standard/Regex.chpl:1157]: The "by calling :proc:`regex.search`" part is my main concern, I don't think the user needs to know necessarily that we're calling the regex's search method.

File: modules/standard/Regex.chpl [hunk 2]
```diff
@@ -1153,28 +1153,52 @@ inline operator :(x: bytes, type t: regex(bytes)) throws {
...
+
 pragma "no doc"
+deprecated "string.search is deprecated, use regex search instead"
 proc string.search(pattern: regex(string)):regexMatch
 {
   return pattern.search(this);
 }
 
+/* Search the receiving bytes for a regular expression already compiled
+   by calling :proc:`regex.search`. Search for matches at any offset.
```
Comment (lydia-duncan) [inline modules/standard/Regex.chpl:1183]: Ditto here

File: modules/standard/Regex.chpl [hunk 3]
```diff
@@ -1243,60 +1277,91 @@ proc bytes.match(pattern: regex(bytes)):regexMatch
    to search for the pattern at any offset.
 
    :arg pattern: the compiled regular expression to match
-   :arg captures: (optional) what to capture from the regular expression. These
-                  should be strings or types that strings can cast to.
-   :returns: an :record:`regexMatch` object representing the offset in the
-             receiving string where a match occurred
+   :returns: true if string starts with :arg pattern:, false otherwise
```
Comment (lydia-duncan) [inline modules/standard/Regex.chpl:1280]: `:arg pattern:` is only useful at the start of a line, where it generates some formatting for it.  I think what you meant is ":arg:`pattern`" but I don't know that that will actually generate a link so it's probably best to surround `pattern` with two backtics.

**Review by lydia-duncan — COMMENTED** [review]
It looks like there's a mention of string.search on line 64 in the file that should be replaced with string.find.

Overall, I mostly was pointing out additional references and things to adjust.  The changes themselves are otherwise pretty good, though there's one or two other things I had some suggestions on.

File: modules/standard/Regex.chpl [hunk 4]
```diff
@@ -1191,17 +1215,22 @@ proc string.search(needle: regex(string), ref captures ...?k):regexMatch
 /* Search the receiving string for a regular expression already compiled
```
[the remark this answers is not in the document]
Comment (slnguyen) [reply modules/standard/Regex.chpl:1215]: What is the argument name are you referring to? Here I intended to deprecate `string.search` and `bytes.search` (which return `regexMatch`) and add `string.find` and `bytes.find` (which return `byteIndex`). Since the `search` function isn't exactly the same as the new `find` function, I didn't mention the new `find` function in the deprecation message.

[the remark this answers is not in the document]
Comment (lydia-duncan) [reply modules/standard/Regex.chpl:1215]: Let me rephrase.  You've got two versions of the same method here, the one on line 1210 and the one on line 1228.  The first one says "the needle argument is deprecated", which to me implies that the user can just use the new argument name and avoid the deprecation warning.  However, the pattern version is also now deprecated.  I think it would be frustrating for a user to make one update and then still have a deprecation warning.  We could just say the function name itself is deprecated for both versions, I was just suggesting a way to preserve the distinction between them.

[the remark this answers is not in the document]
Comment (slnguyen) [reply modules/standard/Regex.chpl:1216]: Okay, got it.

File: modules/standard/Regex.chpl [hunk 5]
```diff
@@ -1215,7 +1216,7 @@ proc string.search(needle: regex(string), ref captures ...?k):regexMatch
...
    by calling :proc:`regex.search`. Search for matches at any offset.
 
-   .. warning:: the search function with a receiving string is deprecated
+   .. warning:: the search function with a receiving string is deprecated,                 use regex search instead
```
Comment (lydia-duncan) [inline modules/standard/Regex.chpl:1219]: This looks a little off, should it be on the following line?

**Review by lydia-duncan — APPROVED** [review]
Other than the minor formatting thing, looks good!  Thanks!
